app/main/views/services.py

```diff
@@ -180,25 +180,40 @@ def update_section(framework_slug, service_id, section_id):
 
     posted_data = section.get_data(request.form)
 
-    try:
-        data_api_client.update_service(
-            service_id,
-            posted_data,
-            current_user.email_address)
-    except HTTPError as e:
-        errors = section.get_error_messages(e.message)
-        if not posted_data.get('serviceName', None):
-            posted_data['serviceName'] = service.get('serviceName', '')
+    errors = None
+    uploaded_documents, document_errors = upload_service_documents(
+        s3.S3(current_app.config['DM_DOCUMENTS_BUCKET']),
+        'documents',
+        current_app.config['DM_ASSETS_URL'],
+        service,
+        request.files,
+        section,
+    )
+    if document_errors:
+        errors = section.get_error_messages(document_errors)
+    else:
+        posted_data.update(uploaded_documents)
+
+    if not errors and section.has_changes_to_save(service, posted_data):
+        try:
+            data_api_client.update_service(
+                service_id,
+                posted_data,
+                current_user.email_address)
+        except HTTPError as e:
+            errors = section.get_error_messages(e.message)
+            if not posted_data.get('serviceName', None):
+                posted_data['serviceName'] = service.get('serviceName', '')
```

Comment:
I know this chunk was in the previous code, but it's not obvious at all what it's meant to be doing. I get that it's for rendering DOS services that don't have a name, but why are we only setting it in the API error case? Maybe an explanatory comment would help 😸

Reply:
If I can work out the point of it I'll add a comment 😛

Reply:
Hmm, looks like I added this back in 2015.  So it must be Important and Useful.

Reply:
And I've remembered what it's for - if someone tries to delete the service name then we want to re-populate the field with the existing service name, rather than just show them an empty box.  It seemed worth it at the time - we really don't want people to have services without a name.

Reply:
Aha - yes, this was more important when the editing was all on one page (as it used to be) rather than one question per page.  The service name used to be the heading for the page, with all the editable fields underneath it. So then if they had deleted their service name and got validation errors the heading disappeared from the page and they didn't know what service they were editing any more.  So this ensured that there was a heading on the page and it was obvious which service they are editing. 

But it still applies for the breadcrumbs now - the final breadcrumb is the service name, so if this has been deleted then the breadcrumb disappears on the validation error page.

Reply:
Makes sense - happy with this comment 👍